src/aiida_quantumespresso/workflows/functions/get_xspectra_structures.py

```diff
@@ -0,0 +1,332 @@
+        point_group = str(analyzer_data.get_pointgroup())
+        output_params['molecule_point_group'] = point_group
+
+        equivalency_dict = {}
+
+        for key in eq_atoms_data:
+            site_symbol = elements[atomic_nos[key]]['symbol']
+            if site_symbol in abs_elements_list:
+                equivalency_dict[f'site_{key}'] = {}
+                atom_no_set = eq_atoms_data[key]
+                equivalency_dict[f'site_{key}']['site_index'] = key
+                equivalency_dict[f'site_{key}']['equivalent_sites_list'] = atom_no_set
+                equivalency_dict[f'site_{key}']['multiplicity'] = len(atom_no_set)
+                equivalency_dict[f'site_{key}']['symbol'] = site_symbol
+
+        output_params['equivalent_sites_data'] = equivalency_dict
+
+        x_pos = []
+        y_pos = []
+        z_pos = []
+        for site in structure.sites:
+            x_pos.append(site.position[0])
+            y_pos.append(site.position[1])
+            z_pos.append(site.position[2])
+
+        # We assume that the Martyna-Tuckerman approach will be used in the core-hole SCF
+        # thus, at a minimum, the new box for the molecule needs to be 2x the extent of the
+        # molecule in x, y, and z. We therefore use the larger of either the MT distance or
+        # `supercell_min_parameter` for each cell parameter.
+        high_x = x_pos[(np.argmax(x_pos))]
+        high_y = y_pos[(np.argmax(y_pos))]
+        high_z = z_pos[(np.argmax(z_pos))]
+
+        low_x = x_pos[(np.argmin(x_pos))]
+        low_y = y_pos[(np.argmin(y_pos))]
+        low_z = z_pos[(np.argmin(z_pos))]
+
+        box_a = max((high_x - low_x) * 2, supercell_min_parameter)
+        box_b = max((high_y - low_y) * 2, supercell_min_parameter)
+        box_c = max((high_z - low_z) * 2, supercell_min_parameter)
+
+        centred_structure = StructureData()
+        centred_structure.set_cell(value=([box_a, 0., 0.], [0., box_b, 0.], [0., 0., box_c]))
+
+        for kind in structure.kinds:
+            centred_structure.append_kind(kind)
+
+        for site in structure.sites:
+            centred_structure.append_site(site)
+
+        # Reset the positions so that the centre of mass is at the centre of the new box
+        centred_structure.reset_sites_positions(centred_positions + (np.array(centred_structure.cell_lengths) / 2))
+
+        output_params['centred_structure_cell_matrix'] = centred_structure.cell
+        output_params['centred_structure_cell_lengths'] = centred_structure.cell_lengths
+        result['centred_structure'] = centred_structure
+
+        for value in equivalency_dict.values():
+            target_site = value['site_index']
+            marked_structure = StructureData()
+            centred_kinds = centred_structure.kinds
+
+            marked_structure.set_cell(centred_structure.cell)
+            for kind in centred_kinds:
+                marked_structure.append_kind(kind)
```

Comment:
Here you copy all kinds to the new structure, thus in the end, you need to validate that all kinds have at least one site. 
For example, for the `CH4` molecule, since there is only one `C` atom. When this site was assigned as `X`,  there is no more site with the kind `C`. Then it shows this error.
```
aiida.common.exceptions.ValidationError: The following kinds are defined, but there are no sites with that kind: ['C']
```
This also applies to larger crystal structures which have only one site for one of the kinds.

Reply:
Yes, that is a fair point. I think it would make sense then to dynamically add `kind`s to the new structure as required, rather than adding an extra check at the end to ensure that no `kind`s are present that do not have a `site`. That seems the simpler option to me at least.